### PR TITLE
Migrate Agreement to FIPS Approved Functions

### PR DESCRIPTION
**Pull request by skmcgrail**
### Before
 
```
Agreement-X25519/AWS-LC time:   [33.262 µs 33.266 µs 33.269 µs]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) low mild
  1 (1.00%) high severe
 
Agreement-P256/AWS-LC   time:   [59.261 µs 59.265 µs 59.268 µs]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
 
Agreement-P384/AWS-LC   time:   [242.20 µs 242.21 µs 242.22 µs]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
```
 
### After
 
```
Agreement-X25519/AWS-LC time:   [33.586 µs 33.590 µs 33.594 µs]
                        change: [+0.9610% +0.9827% +1.0041%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) low mild
  1 (1.00%) high mild
 
Agreement-P256/AWS-LC   time:   [59.677 µs 59.682 µs 59.686 µs]
                        change: [+0.6937% +0.7047% +0.7148%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high severe
 
Agreement-P384/AWS-LC   time:   [242.86 µs 242.88 µs 242.89 µs]
                        change: [+0.2657% +0.2754% +0.2848%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: aws-lc-rs/src/agreement.rs [hunk 1]
```diff
@@ -539,65 +548,79 @@ const MAX_AGREEMENT_SECRET_LEN: usize = ECDH_P521_PRIVATE_KEY_LEN;
...
+    let ec_group = unsafe { ec_group_from_nid(nid)? };
+    let pub_key_point = unsafe { ec_point_from_bytes(&ec_group, peer_pub_key_bytes) }?;
+    let peer_ec_key = unsafe { ec_key_from_public_point(&ec_group, &pub_key_point) }?;
 
-    let priv_group = ConstPointer::new(EC_KEY_get0_group(*priv_ec_key))?;
-    let priv_nid = EC_GROUP_get_curve_name(*priv_group);
+    let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(**priv_key, null_mut()) })?;
 
-    let supported_curves = [NID_X9_62_prime256v1, NID_secp384r1, NID_secp521r1];
-    if !supported_curves.contains(&priv_nid as &i32) {
```
Comment (skmcgrail) [inline aws-lc-rs/src/agreement.rs:556]: From my investigation this check was already covered and duplicative which is why I removed it.

`EVP_PKEY_derive` will call `ECDH_compute_shared_secret` which has the following check: https://github.com/aws/aws-lc/blob/c2c8d66d28bd0421fbe8646e9b1b7e4f94e52757/crypto/fipsmodule/ecdh/ecdh.c#L89-L94

File: aws-lc-rs/src/agreement.rs [hunk 2]
```diff
@@ -226,33 +221,29 @@ impl EphemeralPrivateKey {
...
             AlgorithmID::X25519 => {
-                let mut priv_key = [0u8; X25519_PRIVATE_KEY_LEN];
-                let mut pub_key = [0u8; X25519_PUBLIC_VALUE_LEN];
-                unsafe {
-                    X25519_keypair(pub_key.as_mut_ptr(), priv_key.as_mut_ptr());
-                }
+                let priv_key = generate_x25519()?;
                 Ok(EphemeralPrivateKey {
-                    inner_key: KeyInner::X25519(priv_key),
+                    inner_key: KeyInner::X25519(LcPtr::from(priv_key)),
```
Comment (justsmth) [inline aws-lc-rs/src/agreement.rs:226]: `generate_x25519` and `ec_key_generate` could return `Result<LcPtr<*mut EVP_PKEY>, Unspecified>` -- I don't see anywhere that we call `.detach()` on the returned values.

**Review by justsmth — COMMENTED** [review]
Nice refactoring! Looks good, just a couple of comments.

**Review by justsmth — APPROVED** [review]

File: aws-lc-rs/src/agreement.rs [hunk 3]
```diff
@@ -378,6 +363,36 @@ impl EphemeralPrivateKey {
...
+
+    let ec_group = unsafe { ec_group_from_nid(nid)? };
+    let priv_key = DetachableLcPtr::try_from(priv_key)?;
+
+    let pkey = unsafe { ec::ec_key_from_private(&ec_group.as_const(), &priv_key.as_const())? };
+
+    Ok(pkey)
+}
+
+pub(crate) fn generate_x25519() -> Result<LcPtr<*mut EVP_PKEY>, Unspecified> {
```
Comment (samuel40791765) [inline aws-lc-rs/src/agreement.rs:378]: Not a blocker, but I'm curious if `generate_x25519` can reuse `ec_key_generate` as well, now that we use the same `EVP_PKEY_keygen` APIs in both?

**Review by samuel40791765 — APPROVED** [review]

File: aws-lc-rs/src/agreement.rs (hunk 3 above)
Comment (skmcgrail) [reply aws-lc-rs/src/agreement.rs:378]: ec_key_generate internally also has to set the EC curve parameters on the EVP_KEY, so would still at least need some distinct setup logic. Some of it could potentially be refactored to allow passing a callback function that could set additional properties before the keygen step. I will think it over, and can include it in a separate PR if it makes sense.
